### PR TITLE
UI-Service: Proposal for implementation of modals, 2nd iteration

### PR DESCRIPTION
**Pull request by wanze**
This pull-request contains a refactored implementation of the different modals, based on the discussion in the first pull-request. It also introduces the "Connector" concept which allows UI components to interact between each other (e.g. a clicking on a button triggers to show a modal).

### Examples

Here are some examples to quickly demonstrate the changed interfaces / connector functionality:

**Show an interruptive modal on button click**
```php
global $DIC;
$factory = $DIC->ui()->factory();
$renderer = $DIC->ui()->renderer();
$message = 'Are you sure you want to delete the following items?';
$form_action = $DIC['ilCtrl']->getFormActionByClass('ilsystemstyledocumentationgui');
$modal = $factory->modal()->interruptive('My Title', $message, $form_action)
	->withAffectedItems(array(
		$factory->modal()->interruptiveItem(10, 'Item1'),
		$factory->modal()->interruptiveItem(20, 'Item2'),
		$factory->modal()->interruptiveItem(30, 'Item3'),
	));
$button = $factory->button()->standard('Show Modal', '');
$connection = $factory->connector()->onClick($button, $modal->getShowAction());

$renderer->render([$button, $modal], $connection);
```
**Show a roundtrip modal on button click**
```php
global $DIC;
$factory = $DIC->ui()->factory();
$renderer = $DIC->ui()->renderer();
$action_buttons = array(
	$factory->button()->primary('Primary Action', ''),
	$factory->button()->standard('Secondary Action', ''),
);
$modal = $factory->modal()->roundtrip('My Modal', $factory->legacy('My Content'))
	->withActionButtons($action_buttons);
$button = $factory->button()->standard('Show Modal', '');
$connection = $factory->connector()->onClick($button, $modal->getShowAction());

$renderer->render([$button, $modal], $connection);
```

**Show an interruptive modal with async loaded content**
```php
global $DIC;
$ilCtrl = $DIC['ilCtrl'];
$factory = $DIC->ui()->factory();
$renderer = $DIC->ui()->renderer();
$message = 'Are you sure you want to delete the following items?';
$form_action = $ilCtrl->getFormActionByClass('ilsystemstyledocumentationgui');

// Check if this is the ajax request to deliver the new modal
if (isset($_GET['ajaxRenderModal'])) {
	// Create some random Items
	$items = [
		$factory->modal()->interruptiveItem(10, substr(md5(rand()), 0, 7)),
		$factory->modal()->interruptiveItem(20, substr(md5(rand()), 0, 7)),
	];
	$modal = $factory->modal()->interruptive('My Title', $message, $form_action)
		->withAffectedItems($items);
	echo $renderer->render($modal);
	exit();
}

$modal = $factory->modal()->interruptive('My Title', $message, $form_action);
$button = $factory->button()->standard('Show Modal Async', '');
$ajax_url = $_SERVER['REQUEST_URI'] . '&ajaxRenderModal=1';
$connection = $factory->connector()->onClick($button,$modal->getShowAsyncAction($ajax_url));

$renderer->render([$button, $modal], $connection);
```
**Show a lightbox modal with a single image**
```php
global $DIC;
$factory = $DIC->ui()->factory();
$renderer = $DIC->ui()->renderer();
$button = $factory->button()->standard('Open Modal', '');
$image = $factory->image()->responsive("src/UI/examples/Image/HeaderIconLarge.svg", "Thumbnail Example");
$page = $factory->modal()->lightboxImagePage($image, 'My Title');
$modal = $factory->modal()->lightbox($page);
$connector = $factory->connector()->onClick($button, $modal->getShowAction());

$renderer->render([ $button, $modal ], $connector);
```

### CODE REVIEW

**Comment by klees** [comment]
I'm already looking forward reviewing this =)

**Comment by Amstutz** [comment]
Short note, since there are several key questions about the used patterns open for discussion we will hold a short meeting tomorrow (January 11.) at 14:00. Contact me on skype if you like to join. Further we opened an etherpad to discuss the key aspects of the PR: https://edupad.ch/9uyIzzDLQz.

File: src/UI/Component/Modal/Factory.php [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	 *   effect: >
+	 *      All controls of the original context are inaccessible until the Modal is completed. Upon completion the user
+	 *      returns to the original context.
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: >
+	 *        Due to the heavily disruptive nature of this type of Modal it MUST be restricted to critical situations (e.g. loss of data).
+	 *     2: >
+	 *        All delete-actions are critical situations and MUST be implemented as an Interruptive Modal.
```
Comment (klees) [inline src/UI/Component/Modal/Factory.php:31]: [This](https://github.com/ILIAS-eLearning/ILIAS/pull/379/files/46a48a4fbaade785dd56e40d1e029f03992d74ef#diff-2d456c6ec33f2dfb3c997281c8e23d3fR30) please.

File: src/UI/Component/Modal/Factory.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	 * @return \ILIAS\UI\Component\Modal\RoundTrip
+	 */
+	public function roundtrip($title, $content);
+
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     The Lightbox Modal displays a full of data such as documents, images or a movie inside a Lightbox.
```
Comment (klees) [inline src/UI/Component/Modal/Factory.php:113]: [This](https://github.com/ILIAS-eLearning/ILIAS/pull/379/files/46a48a4fbaade785dd56e40d1e029f03992d74ef#diff-2d456c6ec33f2dfb3c997281c8e23d3fR76) please!

File: src/UI/Component/Modal/Factory.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	 *   composition: >
+	 *     If multiple pages or images are to be displayed in one Lightbox Modal, they can flipped through.
```
Comment (klees) [inline src/UI/Component/Modal/Factory.php:115]: [This](https://github.com/ILIAS-eLearning/ILIAS/pull/379/files/46a48a4fbaade785dd56e40d1e029f03992d74ef#diff-2d456c6ec33f2dfb3c997281c8e23d3fR78) please.

File: src/UI/Component/Modal/Interruptive.php [hunk 4]
```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace ILIAS\UI\Component\Modal;
+
+/**
+ * Interface Interruptive
+ *
+ * @package ILIAS\UI\Component\Modal
+ */
+interface Interruptive extends Modal {
```
Comment (klees) [inline src/UI/Component/Modal/Interruptive.php:10]: Did we consider [this](https://github.com/ILIAS-eLearning/ILIAS/pull/379/files/46a48a4fbaade785dd56e40d1e029f03992d74ef#diff-a44b5109a6891fa08ab7a6cfb2e3b856R21)?

File: src/UI/Component/Modal/InterruptiveItem.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace ILIAS\UI\Component\Modal;
+
+/**
+ * Interface InterruptiveItem
+ *
+ * Represents a item to be displayed inside an interruptive modal
+ */
+interface InterruptiveItem {
```
Comment (klees) [inline src/UI/Component/Modal/InterruptiveItem.php:10]: This is to the interruptive modal like the lightbox page is to the lightbox, right? We should use the same wording then.

File: src/UI/Factory.php [hunk 6]
```diff
@@ -292,4 +292,76 @@ public function legacy($content);
...
+	 *     5: Modals MUST contain a title in the header.
+	 * ---
+	 *
+	 * @return \ILIAS\UI\Component\Modal\Factory
+	 **/
+	public function modal();
+
+
+	/**
+	/**
```
Comment (klees) [inline src/UI/Factory.php:332]: Duplicate.

File: src/UI/Implementation/Component/Modal/Factory.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	 */
+	public function lightbox($pages) {
+		return new Lightbox($pages);
+	}
+
+
+	/**
+	 * @inheritdoc
+	 */
+	public function lightboxImagePage(Component\Image\Image $image, $title, $description = '') {
```
Comment (klees) [inline src/UI/Implementation/Component/Modal/Factory.php:50]: I'm a little irritated about the names with more than one for the pages/items of the modals, we (almost?) do not have these in other places currently. Could we have a factory nesting that allows us to share the lightbox-prefix?

**Review by klees — COMMENTED** [review]
I'm a little lost in the etherpad discussion and don't know how to comment on the implementation atm.

I'm sure Stefan will enlighten me in our meeting =)

File: src/UI/Component/Modal/InterruptiveItem.php (hunk 5 above)
Comment (klees) [reply src/UI/Component/Modal/InterruptiveItem.php:10]: Just understood that this is not the case.

File: src/UI/Implementation/Component/Modal/Factory.php (hunk 7 above)
Comment (wanze) [reply src/UI/Implementation/Component/Modal/Factory.php:50]: I'm not sure I fully understand, do you mean a factory for generating the different lightbox pages, e.g.
```
uiFactory()->modal()->lightboxPage()->image()
uiFactory()->modal()->lightboxPage()->video()
```
etc.

File: src/UI/Component/Modal/Interruptive.php (hunk 4 above)
Comment (wanze) [reply src/UI/Component/Modal/Interruptive.php:10]: The action button is now mandatory and the interruptive modal takes a form action where the action button will send the POST request, optionally containing the IDs of the affected items.
The label of the action button is injected as language key with the constructor: https://github.com/ILIAS-eLearning/ILIAS/pull/404/files#diff-4a17a81d57879772cca64c5300e12065R43
However, this is currently not exposed to the factory: What's the strategy here?

File: src/UI/Component/Modal/Factory.php (hunk 3 above)
Comment (wanze) [reply src/UI/Component/Modal/Factory.php:115]: @Amstutz @klees I'm really not sure what I should write in composition now - can you give me a copy paste sentence please?

File: src/UI/Implementation/Component/Modal/Factory.php (hunk 7 above)
Comment (wanze) [reply src/UI/Implementation/Component/Modal/Factory.php:50]: > Could we have a factory nesting that allows us to share the lightbox-prefix?

@klees Does my pseudocode snippet above the thing you imagined?  :)

**Comment by wanze** [comment]
Closed, a new pull request was opened: 
https://github.com/ILIAS-eLearning/ILIAS/pull/421

**Pull request closed without merging**